lib/osutil/osutil.go

```diff
@@ -22,37 +22,78 @@ import (
 // often enough that there is any contention on this lock.
 var renameLock = sync.NewMutex()
 
-// TryRename renames a file, leaving source file intact in case of failure.
+// RenameOrCopy renames a file, leaving source file intact in case of failure.
 // Tries hard to succeed on various systems by temporarily tweaking directory
 // permissions and removing the destination file when necessary.
-func TryRename(filesystem fs.Filesystem, from, to string) error {
+func RenameOrCopy(src, dst fs.Filesystem, from, to string) error {
 	renameLock.Lock()
 	defer renameLock.Unlock()
 
-	return withPreparedTarget(filesystem, from, to, func() error {
-		return filesystem.Rename(from, to)
-	})
-}
+	return withPreparedTarget(dst, from, to, func() error {
+		// Optimisation 1
+		if src.Type() == dst.Type() && src.URI() == dst.URI() {
+			return src.Rename(from, to)
+		}
 
-// Rename moves a temporary file to its final place.
-// Will make sure to delete the from file if the operation fails, so use only
-// for situations like committing a temp file to its final location.
-// Tries hard to succeed on various systems by temporarily tweaking directory
-// permissions and removing the destination file when necessary.
-func Rename(filesystem fs.Filesystem, from, to string) error {
-	// Don't leave a dangling temp file in case of rename error
-	if !(runtime.GOOS == "windows" && strings.EqualFold(from, to)) {
-		defer filesystem.Remove(from)
-	}
-	return TryRename(filesystem, from, to)
+		// "Optimisation" 2
+		// Try to find a common prefix between the two filesystems, use that as the base for the new one
+		// and try a rename.
+		if src.Type() == dst.Type() {
+			srcURI := []byte(src.URI())
+			dstURI := []byte(dst.URI())
+			smaller := len(srcURI)
+			if len(srcURI) > len(dstURI) {
+				smaller = len(dstURI)
+			}
+
+			commonBytes := make([]byte, 0, smaller)
+
+			for i := 0; i < smaller; i++ {
+				if srcURI[i] != dstURI[i] {
+					break
+				}
+				commonBytes = append(commonBytes, srcURI[i])
+			}
```

Comment:
Just something I noticed: `commonBytes` must not represent a valid path , e.g. `/real/fake-first` and `/real/fake-second` results in `/real/fake`. Any operation on this filesystem should fail, so it probably will still do the right thing (not do anything). And bytes might be faster than a solution using `filepath.Split` or `strings.Split(path, fs.PathSeparator)` - or not.